src/core/control/Control.cpp

```diff
@@ -2403,6 +2406,62 @@ auto Control::openFile(fs::path filepath, int scrollToPage, bool forceOpen) -> b
     return true;
 }
 
+auto Control::changePdfBackground() -> bool {
+
+    LoadHandler loadHandler;
+
+    const fs::path currentFilepath = ((*this).doc)->getFilepath();
+    const fs::path currentPdfFilepath = ((*this).doc)->getPdfFilepath();
+
+    auto pageNr = getCurrentPageNo();
+    PageRef page = this->doc->getPage(pageNr);
+    // Layer* layer = page->getSelectedLayer();
+
+    if (currentPdfFilepath.empty()) {
+        string msg = FS(_F("Error in changing the background: please open a pdf first."));
+        XojMsgBox::showErrorToUser(getGtkWindow(), msg);
+        return false;
+    }
+
+    if (this->undoRedo->isChanged()) {
+        string msg = FS(_F("Error in changing the background: please save the recent changes first."));
+        XojMsgBox::showErrorToUser(getGtkWindow(), msg);
+        return false;
+    }
+
+    if (currentFilepath.empty()) {
+        string msg = FS(_F("Error in changing the background: please save the document first."));
+        XojMsgBox::showErrorToUser(getGtkWindow(), msg);
+        return false;
+    }
+
+    Document* loadedDocument = loadHandler.loadDocument(currentFilepath);
+
+    bool attachToDocument = false;
+    XojOpenDlg dlg(getGtkWindow(), this->settings);
+    auto pdfFilename = dlg.showOpenDialog(true, attachToDocument);
+    if (!pdfFilename.empty()) {
+        loadHandler.setPdfReplacement(pdfFilename, attachToDocument);
+        loadedDocument = loadHandler.loadDocument(currentFilepath);
+        if (!loadedDocument) {
+            string msg = FS(_F("Error opening file \"{1}\"") % currentFilepath.u8string()) + "\n" +
+                         loadHandler.getLastError();
+            XojMsgBox::showErrorToUser(getGtkWindow(), msg);
+            return false;
+        }
+    }
+
+    this->closeDocument();
+
+    this->doc->lock();
+    this->doc->clearDocument();
+    *this->doc = *loadedDocument;
+    this->doc->unlock();
```

Comment:
Probably something like this:
<details>
<summary>Patch to the existing PR</summary>

```diff
diff --git a/src/core/control/Control.cpp b/src/core/control/Control.cpp
index 45306ce8e..281c2f07f 100644
--- a/src/core/control/Control.cpp
+++ b/src/core/control/Control.cpp
@@ -2435,30 +2435,24 @@ auto Control::changePdfBackground() -> bool {
         return false;
     }
 
-    Document* loadedDocument = loadHandler.loadDocument(currentFilepath);
-
     bool attachToDocument = false;
     XojOpenDlg dlg(getGtkWindow(), this->settings);
     auto pdfFilename = dlg.showOpenDialog(true, attachToDocument);
-    if (!pdfFilename.empty()) {
-        loadHandler.setPdfReplacement(pdfFilename, attachToDocument);
-        loadedDocument = loadHandler.loadDocument(currentFilepath);
-        if (!loadedDocument) {
-            string msg = FS(_F("Error opening file \"{1}\"") % currentFilepath.u8string()) + "\n" +
-                         loadHandler.getLastError();
-            XojMsgBox::showErrorToUser(getGtkWindow(), msg);
-            return false;
-        }
-    }
-
-    this->closeDocument();
 
-    this->doc->lock();
-    this->doc->clearDocument();
-    *this->doc = *loadedDocument;
-    this->doc->unlock();
+    // To prevent render jobs from running while we are changing the document,
+    // we need to lock the scheduler.
+    getScheduler()->lock();
+    bool success = this->doc->readPdf(pdfFilename, false, attachToDocument);
+    if (!success) {
+        getScheduler()->unlock();
+        g_warning("Failed to read new PDF document.");
+    } else {
+        fileLoaded(-1);
+        getScheduler()->unlock();
+        // Pretend the whole document has changed to force re-render.
+        this->win->getXournal()->documentChanged(DOCUMENT_CHANGE_COMPLETE);
+    }
 
-    fileLoaded(-1);
     return true;
 }

```
</details>

The last call to `his->win->getXournal()->documentChanged(DOCUMENT_CHANGE_COMPLETE)` is a hack to invalidate the PdfCache, which will hold a reference to the poppler PDF document. I think the old poppler PDF document will be destroyed after we load the new PDF, hence the need to reset the cache.

I haven't tested this though, so not sure if it will work. This is more complicated code, but it will also allow a UndoAction to be defined (if that is something you'd like to address in this PR).

Reply:
Thank you for the patch, the code compiles without errors.
I wasn't thinking about an UndoAction for this PR.